File: src/main/java/skills/XCv.scala
package skills

object XCv {

  private class Thou extends XpCurve {
    override def getLevel(xp: Int): Int = {
      xp/1000
    }

    override def xpForLevel(level: Int):Int = {
      level*1000
    }
  }
  private class Cube() extends XpCurve {
    override def getLevel(xp: Int): Int = {
      Math.cbrt(xp).toInt;
    }

    override def xpForLevel(level: Int):Int = {
      level * level * level
    }
  }
  // the factory method
  def apply(): XpCurve = {
    new Cube;

  }
}

File: src/main/java/skills/XCv.java
package skills;

import org.jetbrains.annotations.Contract;

/**
 * Created by nichita on 12/01/17.
 */
public class XCv {

    private static class Thou implements XpCurve {
        @Override
        public int getLevel(int xp) {
            return xp/1000;
        }

        @Override
        public int xpForLevel(int level) {
            return level*1000;
        }
    }

    private static class Cube implements XpCurve {
        @Override
        public int getLevel(int xp) {
            return (int)Math.cbrt(xp);
        }

        @Override
        public int xpForLevel(int level) {
            return level * level * level;
        }
    }

    // the factory method
    @Contract(" -> !null")
    public static XpCurve apply(){
        return new Cube();

    }
}
